test/k8sT/Policies.go

```diff
@@ -1115,6 +1117,191 @@ var _ = Describe("K8sPolicyTest", func() {
+			var (
+				outsideNodeName, outsideIP string // k8s3 node (doesn't have agent running)
+
+				backendPod   v1.Pod // The pod that k8s3 node is hitting
+				backendPodIP net.IP
+
+				hostNodeName       string // Node that backendPod ends up on
+				hostIPOfBackendPod net.IP
+
+				policyVerdictAllowRegex, policyVerdictDenyRegex *regexp.Regexp
+			)
+
+			BeforeAll(func() {
+				RedeployCiliumWithMerge(kubectl, ciliumFilename, daemonCfg,
+					map[string]string{
+						"global.tunnel":                 "disabled",
+						"global.autoDirectNodeRoutes":   "true",
+						"global.endpointRoutes.enabled": "true",
+
+						// Needed because of
+						// https://github.com/cilium/cilium/issues/12141
+						"global.masquerade":   "false",
+						"global.devices":      "",
+						"global.hostFirewall": "disabled",
+					})
+
+				outsideNodeName, outsideIP = kubectl.GetNodeInfo(helpers.GetNodeWithoutCilium())
+				Expect(outsideNodeName).ShouldNot(BeEmpty())
+				Expect(outsideIP).ShouldNot(BeEmpty())
+
+				var demoPods v1.PodList
+				kubectl.GetPods(testNamespace, fmt.Sprintf("-l %s", testDS)).Unmarshal(&demoPods)
+				Expect(demoPods.Items).Should(HaveLen(2))
+
+				backendPod = demoPods.Items[0] // We'll take the first one; doesn't matter
+				backendPodIP = net.ParseIP(backendPod.Status.PodIP)
+				hostIPOfBackendPod = net.ParseIP(backendPod.Status.HostIP)
+				hostNodeName = backendPod.Spec.NodeName // Save the name of node backend pod is on
+
+				// Add the route on the outside node to the backend pod IP
+				// directly.
+				res, err := kubectl.ExecInHostNetNS(context.TODO(), outsideNodeName,
+					helpers.IPAddRoute(backendPodIP, hostIPOfBackendPod, true))
+				ExpectWithOffset(1, res).Should(helpers.CMDSuccess())
+				ExpectWithOffset(1, err).ShouldNot(HaveOccurred(),
+					"Cannot exec in outside node %s", outsideNodeName)
+
+				policyVerdictAllowRegex = regexp.MustCompile(
+					fmt.Sprintf("Policy verdict log: .+action allow.+%s:[0-9]+ -> %s:80 tcp SYN",
+						outsideIP,
+						backendPodIP.String()))
+				policyVerdictDenyRegex = regexp.MustCompile(
+					fmt.Sprintf("Policy verdict log: .+action deny.+%s:[0-9]+ -> %s:80 tcp SYN",
+						outsideIP,
+						backendPodIP.String()))
+			})
+
+			AfterAll(func() {
+				// Remove the route on the outside node.
+				kubectl.ExecInHostNetNS(context.TODO(), outsideNodeName,
+					helpers.IPDelRoute(backendPodIP, hostIPOfBackendPod))
+
+				// Revert Cilium installation back to before this Context.
+				By("Redeploying Cilium with default configuration")
+				RedeployCilium(kubectl, ciliumFilename, daemonCfg)
+			})
+
+			testConnectivity := func(expectSuccess bool) int {
+				action := "allowed"
+				if !expectSuccess {
+					action = "denied"
+				}
+				By("Testing that connectivity from outside node is %s", action)
+
+				// This Eventually waits for the traffic to eventually succeed
+				// or fail (as we wait for the policy to take effect).
```

Comment:
Yeah, I think you're right in that the `Eventually` may not be required. I thought that the test was flakey in CI, and I wasn't seeing any flakes locally. 

It turns out the test isn't flakey from what I can tell. It was due to using per-endpoint routes in this test context..The demo pods (test pods) were never deleted and redeployed to run under this configuration. I believe this was causing the tests to fail in CI. I wasn't seeing it locally because I had commented out the top-level deploy of CIlium for the sake of speed, so the only deploy of Cilium happening was in this test's context. In other words, the demo pods' plumbing was in-sync with what Cilium was expecting when I was running locally, but not in CI.

This should be fixed now that I've removed configuring per-endpoint routes.